lgc/builder/InOutBuilder.cpp

```diff
@@ -767,6 +767,88 @@ Instruction *InOutBuilder::CreateWriteXfbOutput(Value *valueToWrite, bool isBuil
   return CreateNamedCall(instName, getVoidTy(), args, {});
 }
 
+// Create a read of barycoord input value.
+// The type of the returned value is the fixed type of the specified built-in (see BuiltInDefs.h),
+//
+// @param builtIn : Built-in kind, BuiltInBaryCoord or BuiltInBaryCoordNoPerspKHR
+// @param inputInfo : Extra input info
+// @param auxInterpValue : Auxiliary value of interpolation
+// @param instName : Name to give instruction(s)
+Value *InOutBuilder::CreateReadBaryCoord(BuiltInKind builtIn, InOutInfo inputInfo, llvm::Value* auxInterpValue,
+                                         const llvm::Twine& instName) {
+   assert(builtIn == lgc::BuiltInBaryCoord || builtIn == lgc::BuiltInBaryCoordNoPerspKHR);
+
+   if (getPipelineState()->getOptions().enableInterpModePatch && !auxInterpValue &&
+       inputInfo.getInterpLoc() != InOutInfo::InterpLocCentroid) {
+     auxInterpValue = readBuiltIn(false, BuiltInSampleId, {}, auxInterpValue, nullptr, "");
+     inputInfo.setInterpLoc(InOutInfo::InterpLocSample);
+   }
+
+   // Get the built-in type.
+   Type *resultTy = getBuiltInTy(builtIn, inputInfo);
+   bool isPersp = builtIn == lgc::BuiltInBaryCoord;
+   Value *evalArg = nullptr;
+   // Adjust ijCoord
+   switch (inputInfo.getInterpLoc())
+   {
+   case InOutInfo::InterpLocCentroid: {
+     std::string evalInstName = lgcName::InputImportBuiltIn;
+     if (isPersp) {
+       evalInstName += "InterpPerspCentroid";
+       evalArg = getInt32(BuiltInInterpPerspCentroid);
+     } else {
+       evalInstName += "InterpLinearCentroid";
+       evalArg = getInt32(BuiltInInterpLinearCentroid);
+     }
+     auxInterpValue =
+         CreateNamedCall(evalInstName, FixedVectorType::get(getFloatTy(), 2), {evalArg}, Attribute::ReadOnly);
+     break;
+   }
+   case InOutInfo::InterpLocSample: {
+     assert(auxInterpValue != nullptr);
+     auxInterpValue = readBuiltIn(false, BuiltInSamplePosOffset, {}, auxInterpValue, nullptr, "");
+     if (isPersp)
+       auxInterpValue = evalIjOffsetSmooth(auxInterpValue);
+     else
+       auxInterpValue = evalIjOffsetNoPersp(auxInterpValue);
+     break;
+   }
+   default: {
+     // When calling InterpolateAtOffset, auxInterpValue != nullptr
```

Comment:
Using the `default` in this way is confusing. Better to explicitly have distinct cases for `InterpLocCenter` and `InterpLocExplicit`.